httpx/client.py

```diff
@@ -176,6 +179,13 @@ def merge_cookies(
             response = await self.dispatch.send(
                 request, verify=verify, cert=cert, timeout=timeout
             )
+            # TODO: requests also checks for number of 401 calls < 2
+            if response.is_client_error and response.expects_digest_auth:
+                request = self.auth(request, response)  # TODO: remove dependency on self.auth
```

Comment:
We're currently following request's design of using a `BaseAuth` class and, for now, a single concrete implementation with a `__call__` method that injects the authorization header into the request. However this does not quite fit the Digest authentication flow where the authorization header depends on the `www-authenticate` header of the response, which breaks the abstraction.

To get around it temporarily I'm assuming the user creates an `HTTPDigestAuth` object manually [as described in the requests docs](https://2.python-requests.org//en/master/user/authentication/#digest-authentication) and storing it for later use in `self.auth` but ideally we'd redesign the abstraction so it takes the request and optionally a response.

Requests gets around it by using a [`register_hook` callback](https://github.com/kennethreitz/requests/blob/master/requests/auth.py#L292) in the `HTTPDigestAuth` which takes the response and does the tasks we're doing in the `client.send_handling_redirects`.

Thoughts?

Reply:
Not sure on first sight, but this is a *really* useful point of reference wrt. what API we need to support. Not comfortable jumping on any particular answer quickly here, and it'd push me a bit towards thinking that we *might* want to rethink the auth API.

Reply:
Also nudges me a bit more in the direction of thinking that DigestAuth should exist as a third party auth, rather than us building it in. (I don't mind us working through it with a PR, but we may well eventually want to consider pushing it a seperate package.)

Reply:
I tend to agree, the Digest flow is complex enough to grant its own package, in a similar vein to [`requests-oauthlib`](https://github.com/requests/requests-oauthlib).

I think I'll start a separate issue to discuss how to best approach it design-wise and close this PR.

Reply:
Sure thing. Also very welcome to work through it against this pull request (as a useful place for discussion, and to validate test cases against it) and only start pulling it out into a package once you've got something sufficiently mature. I don't mind either ways.